textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java

```diff
@@ -0,0 +1,149 @@
+package edu.uci.ics.textdb.dataflow.queryrewriter;
+
+import edu.uci.ics.textdb.dataflow.dictionarymatcher.Dictionary;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by shiladityasen on 4/30/16.
+ * Brute force algorithm to take input a phrase string and output an array of strings as rewritten phrases
+ * A phrase is a string that may comprise of multiple space separated strings called terms
+ * Algorithm tries to find a valid prefix of a term and parses the rest of the term recursively
+ * After recursive call returns with a set of parsed strings, prefix is appended in front of each string
+ *
+ * For example -> ["newyorkcity"] -> ["new york city", "newyorkcity"]
+ * Here, "new york city" is a returned phrase because "new", "york" and "city" are valid words in the word base.
+ */
+public class FuzzyTokenizer
+{
+    //Data members
+    private static Dictionary wordBase;
+    private String phrase;
+    private static final String wordBaseSourceFilePath = "/queryrewriter/wordsEn.txt";
+
+    static {
+        try {
+            wordBase = new Dictionary(wordBaseSourceFilePath);
+        }
+        catch(IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Parameterized constructor requires phrase string that is to be tokenized
+     * Creates wordBase object containing knowledge base of words in file wordsEn.txt
+     * @param phrase
+     * @throws IOException
+     */
+    public FuzzyTokenizer(String phrase) {
+        this.phrase = phrase;
+    }
+
+    /**
+     * When called returns array of rewritten phrases
+     * Splits phrase string into delimited (by default, it is delimited using a space) terms
+     * For each term performs fuzzy tokenization to create corresponding lists of rewritten terms
+     * Calls method rewriteTerm to perform tokenization over single term
+     * Performs a cross product concatenation of each rewritten term list and returns final String array
+     * Calls method crossCatenate to perform required concatenation of multiple lists
+     * @return
+     */
+    public List<String> getFuzzyTokens() {
+        String[] queryWords = phrase.split("\\s+");
+        List< List<String> > allWordsList = new ArrayList< List <String> >();
+        for(String word : queryWords)
+            allWordsList.add(rewriteTerm(word));
+        return crossCatenate(allWordsList);
+    }
+
+    /**
+     * Wrapper over main recursive method, rewrite which performs the fuzzy tokenization of a term
+     * Ensures that original phrase term is also included in returned rewritten terms list
+     * Is necessary if original phrase term is not a valid word according to word knowledge base
+     * @param term
+     * @return
+     */
+    private List<String> rewriteTerm(String term) {
+        List<String> queryWordList = rewrite(term);
+        if(! term.equals(queryWordList.get(queryWordList.size()-1)))
```

Comment:
Not quite follow this statement.  Can you add a comment in the codebase to explain?